core/src/test/scala/org/apache/spark/scheduler/BlacklistTrackerSuite.scala

```diff
@@ -0,0 +1,503 @@
+      .set(config.BLACKLIST_ENABLED.key, "true")
+    val scheduler = mockTaskSchedWithConf(conf)
+    // Task 1 failed on executor 1
+    blacklistTracker = new BlacklistTracker(conf, clock)
+    val taskSet = FakeTask.createTaskSet(10)
+    val tsm = new TaskSetManager(scheduler, Some(blacklistTracker), taskSet, 4, clock)
+    tsm.updateBlacklistForFailedTask("hostA", "1", 0)
+    for {
+      executor <- (1 to 4).map(_.toString)
+      index <- 0 until 10
+    } {
+      val exp = (executor == "1"  && index == 0)
+      assert(tsm.isExecutorBlacklistedForTask(executor, index) === exp)
+    }
+    assert(blacklistTracker.nodeBlacklist() === Set())
+    assertEquivalentToSet(blacklistTracker.isNodeBlacklisted(_), Set())
+    assertEquivalentToSet(tsm.isNodeBlacklistedForTaskSet, Set())
+    assertEquivalentToSet(tsm.isExecutorBlacklistedForTaskSet, Set())
+
+    // Task 1 & 2 failed on both executor 1 & 2, so we blacklist all executors on that host,
+    // for all tasks for the stage.  Note the api expects multiple checks for each type of
+    // blacklist -- this actually fits naturally with its use in the scheduler
+    tsm.updateBlacklistForFailedTask("hostA", "1", 1)
+    tsm.updateBlacklistForFailedTask("hostA", "2", 0)
+    tsm.updateBlacklistForFailedTask("hostA", "2", 1)
+    // we don't explicitly return the executors in hostA here, but that is OK
+    for {
+      executor <- (1 to 4).map(_.toString)
+      index <- 0 until 10
+    } {
+      withClue(s"exec = $executor; index = $index") {
+        val badExec = (executor == "1" || executor == "2")
+        val badPart = (index == 0 || index == 1)
+        val taskExp = (badExec && badPart)
+        assert(
+          tsm.isExecutorBlacklistedForTask(executor, index) === taskExp)
+        val executorExp = badExec
+        assert(tsm.isExecutorBlacklistedForTaskSet(executor) === executorExp)
+      }
+    }
+    assertEquivalentToSet(tsm.isNodeBlacklistedForTaskSet, Set("hostA"))
+    // we dont' blacklist the nodes or executors till the stages complete
+    assert(blacklistTracker.nodeBlacklist() === Set())
+    assertEquivalentToSet(blacklistTracker.isNodeBlacklisted(_), Set())
+    assertEquivalentToSet(blacklistTracker.isExecutorBlacklisted(_), Set())
+
+    // when the stage completes successfully, now there is sufficient evidence we've got
+    // bad executors and node
+    blacklistTracker.updateBlacklistForSuccessfulTaskSet(0, 0, tsm.execToFailures)
+    assert(blacklistTracker.nodeBlacklist() === Set("hostA"))
+    assertEquivalentToSet(blacklistTracker.isNodeBlacklisted(_), Set("hostA"))
+    assertEquivalentToSet(blacklistTracker.isExecutorBlacklisted(_), Set("1", "2"))
+
+    clock.advance(blacklistTracker.BLACKLIST_TIMEOUT_MILLIS + 1)
+    blacklistTracker.applyBlacklistTimeout()
+    assert(blacklistTracker.nodeBlacklist() === Set())
+    assertEquivalentToSet(blacklistTracker.isNodeBlacklisted(_), Set())
```

Comment:
It would be great to add one more test for the timeout, that makes sure that if multiple executors (or hosts) are blacklisted, and the timeout for just one expires, only that one gets un-blacklisted